Sample Finder pathways form samples/sources/registry grids via FindDerivativesButton

cnathe opened the pull request:
#### Rationale
This story adds a new "Find Derivatives" button to several of the samples/sources/registry app grids so that users can easily take the filters they have applied to a given grid and translate that over to the sample finder. This PR adds the new FindDerivativesButton which is added to the grids and on click will take any provided base filters, the view filters, and the user defined filters and add them to session storage as a sample finder search result to be viewed by the user.

#### Related Pull Requests
* https://github.com/LabKey/labkey-ui-components/pull/971
* https://github.com/LabKey/sampleManagement/pull/1250
* https://github.com/LabKey/biologics/pull/1631
* https://github.com/LabKey/inventory/pull/556

#### Changes
* FindDerivativesButton added to samples/sources/registry grids
  * take the model view and user filters and create the sample finder sessionStorage object
  * allow passthrough of base/context filter to be applied for certain scenarios
* fix for "Showing all samples with ... parents" filter to use getExpDescendantOfFilter()
* add QueryModel helper to get just the view filters from the queryInfo / viewName

cnathe committed:
Initial version of FindDerivativesButton to add to sample/sources/registry grids
- when clicked, it will take the model view and user filters and create the sample finder sessionStorage object
- add QueryModel helper to get just the view filters from the queryInfo / viewName
- rename getLocalStorageKey to getSampleFinderLocalStorageKey

cnathe committed:
v2.219.0-fb-sampleFinderPaths.0

cnathe committed:
Merge branch 'develop' into fb_sampleFinderPaths

cnathe committed:
Merge branch 'develop' into fb_sampleFinderPaths

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.220.0-fb-sampleFinderPaths.0

cnathe committed:
SampleFinderSection.tsx to load view from URL param on init

cnathe committed:
location null check

cnathe committed:
v2.220.0-fb-sampleFinderPaths.1

cnathe committed:
FindDerivativesButton.tsx to allow pass through of base/context filter

cnathe committed:
v2.220.0-fb-sampleFinderPaths.2

cnathe committed:
FindDerivativesButton.tsx to allow pass through of base/context filter (part 2)

cnathe committed:
v2.220.0-fb-sampleFinderPaths.3

cnathe committed:
metricFeatureArea and call to incrementClientSideMetricCount

cnathe committed:
v2.220.0-fb-sampleFinderPaths.4

cnathe committed:
Merge branch 'develop' into fb_sampleFinderPaths

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.222.0-fb-sampleFinderPaths.0

cnathe committed:
Sample Finder fix for "Showing all samples with ... parents" filter to use getExpDescendantOfFilter()

cnathe committed:
v2.222.0-fb-sampleFinderPaths.1

cnathe committed:
Sample Finder fix for "Showing all samples with ... parents" filter to use getExpDescendantOfFilter() - part 2

cnathe committed:
v2.222.0-fb-sampleFinderPaths.2

cnathe committed:
update release notes

cnathe committed:
Sample Finder to not apply and instead display error alert for invalid filter values

cnathe committed:
v2.222.0-fb-sampleFinderPaths.3

cnathe committed:
revert previous commit for MVFK Sample Finder filter handling

cnathe committed:
disable FindDerivativesButton.tsx if an invalid filter is added to grid (i.e. a filter on a MVFK field that doesn't work in Sample Finder)

cnathe committed:
v2.222.0-fb-sampleFinderPaths.4

cnathe committed:
update FindDerivativesButton.tsx disable message to include field caption

cnathe committed:
v2.222.0-fb-sampleFinderPaths.5

cnathe committed:
Merge branch 'develop' into fb_sampleFinderPaths

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md

cnathe committed:
v2.222.1-fb-sampleFinderPaths.0

cnathe committed:
jest tests (part 1)

XingY commented on packages/components/src/internal/components/search/utils.ts:
```diff
@@ -718,6 +727,17 @@ export function isValidFilterFieldExcludeLookups(
     return !field.isLookup();
 }
 
+export function isValidFilterFieldSampleFinder(
```
should the check of multiValue be done in isValidFilterField directly? 

Also, I notice that "Units" is filterable (via column header) on samples grid, but the filtering is not doing the right thing. We should perhaps disable filtering on Unit there as well.

cnathe replied:
General grid filtering does work for `multiValue` fields, so I don't think we want this check in `isValidFilterField`

For the units column grid filtering, are you seeing errors or just getting unexpected results in the grid? This seems to be working for me locally.

XingY replied:
i am not seeing errors but I'm seeing unexpected results. If a sample doesn't have a unit (item.units is blank), the sample type's unit will be displayed. However, filtering filters on item.units, instead of the displayed value.

cnathe replied:
Ah, I see now. Yeah that isn't behaving as I would expect. I agree we should just turn off filtering for the Units field for now. Done.

cnathe committed:
F

cnathe committed:
v2.222.1-fb-sampleFinderPaths.1

cnathe committed:
jest tests for FindDerivativesButton.tsx

cnathe committed:
v2.222.1-fb-sampleFinderPaths.2

cnathe committed:
selenium test fix - add null check for model.queryInfo before getting viewFilters

cnathe committed:
v2.222.1-fb-sampleFinderPaths.3

XingY commented on packages/components/src/internal/components/search/utils.spec.ts:
```diff
@@ -1260,6 +1301,25 @@ describe('isValidFilterField', () => {
+        const queryInfo = QueryInfo.create({
+            schemaName: 'test',
+            name: 'query',
+            supportGroupConcatSubSelect: true,
```
Can consider adding a test for supportGroupConcatSubSelect=false as well.

cnathe replied:
done

XingY commented on packages/components/src/internal/components/search/utils.ts:
```diff
@@ -179,7 +179,16 @@ export function getSampleFinderCommonConfigs(
         } else {
-            baseFilters.push(Filter.create(cardColumnName + '/Name', null, Filter.Types.NONBLANK));
+            const pkColName = 'Name';
+            const filter = getExpDescendantOfFilter(schemaQuery, [{
```
Curious to know why the previous query was wrong.

XingY replied:
Also, does this mean existing saved view with the old filter is broken?

cnathe replied:
I think we originally added this during a very early phase of the sample finder work, but it wasn't actually possible to create this type of filter directly from the Sample Finder UI (as far as I could tell), so the code became unused. With this code path, you can get this filter again but I wasn't seeing the correct results. When I compared the network tab filtering for the previous code and what I got when I manually added a "Name is not null" filter, it was different so I changed it to match that by using the `getExpDescendantOfFilter` helper. Let me know if you see any issues in manual testing or unexpected results in the grids.

XingY commented on packages/components/src/internal/components/search/FindDerivativesButton.tsx:
```diff
@@ -0,0 +1,156 @@
+        fieldKey: column?.fieldKey ?? colName,
+        fieldCaption: column?.caption ?? colName,
+        filter,
+        jsonType: column.isLookup() ? 'string' : column?.jsonType ?? 'string', // deferring to 'string' for lookups since lookup display columns default to text fields
```
should we use column.jsonType or column.displayFieldJsonType here? Sample Finder uses displayFieldJsonType

cnathe replied:
Perfect, yes I'll switch to `column.displayFieldJsonType`. I didn't see that in the props before. Thanks.

XingY commented on packages/components/src/internal/components/search/FindDerivativesButton.tsx:
```diff
@@ -0,0 +1,156 @@
+    if (baseModel && baseFilter) {
+        filterProps.push({
+            schemaQuery: baseModel.schemaQuery,
+            filterArray: [getFieldFilter(baseModel, baseFilter[0])],
```
why only baseFilter[0]?

cnathe replied:
The use case here is when the LKSM source > samples usage passes through the "Source ID = X" baseFilter, which will always be passing only a single filter. But I can update this to generalize for a case that might pass multiple. Done.

XingY commented on packages/components/src/internal/components/search/FindDerivativesButton.tsx:
```diff
@@ -0,0 +1,156 @@
+        filterProps.push({
+            schemaQuery: baseModel.schemaQuery,
+            filterArray: [getFieldFilter(baseModel, baseFilter[0])],
+            entityDataType:
```
parentEntityDataTypes are passed in from Biologics and LKSM and are not the static const version of DataClassDataType and SampleTypeDataType. 

Also, why is there a need to determine entityDataType if it's passed in as param already?

cnathe replied:
the `entityDataType` prop that is passed in is for the main set of filters on the given grid. In this baseFilter case, that can be different then what we want to use. For example the LKSM Source > Samples grids where the grid filters use the `SampleTypeDataType` but the baseFilter is for the source and uses `DataClassDataType`.

cnathe replied:
ah, I see where it matters not. I'll change it so that the usage needs to pass in the baseEntityDataType so we get the labels right in the SF cards

cnathe committed:
FindDerivativesButton.tsx fixes
- use column.displayFieldJsonType for lookups
- take in baseEntityDataType as prop instead of using DataClassDataType and directly

cnathe committed:
v2.222.1-fb-sampleFinderPaths.4

cnathe committed:
FindSamplesByIdsPageBase.tsx update to use Name IN clause instead of RowId IN clause

cnathe committed:
v2.222.1-fb-sampleFinderPaths.5

cnathe committed:
use filter columnName as fieldKey in SF card

cnathe committed:
v2.222.1-fb-sampleFinderPaths.6

cnathe committed:
Merge branch 'develop' into fb_sampleFinderPaths

# Conflicts:
#	packages/components/package.json
#	packages/components/releaseNotes/components.md
#	packages/components/src/internal/components/search/SampleFinderSection.tsx
#	packages/components/src/internal/components/search/utils.ts

cnathe committed:
yarn run lint-branch-fix

cnathe committed:
Merge branch 'develop' into fb_sampleFinderPaths

# Conflicts:
#	packages/components/releaseNotes/components.md

cnathe committed:
Update release notes with version number and release date

cnathe committed:
v2.225.0
